File: scripts/lib/daily-reflection.ts
/**
 * Daily reflection creation and update for Obsidian vault.
 *
 * Deterministic, template-based operations — no AI required.
 * Creates/updates daily reflection notes and inserts blog post links.
 *
 * All pure functions operate on strings; I/O functions handle the filesystem.
 *
 * @module daily-reflection
 */

import fs from "node:fs";
import path from "node:path";

import type { BlogSeriesConfig } from "./blog-series-config.ts";
import {
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
} from "./types.ts";

// --- Constants ---

const EMBED_SECTION_HEADERS = [
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
] as const;

// --- Result Types ---

export interface EnsureReflectionResult {
  readonly created: boolean;
  readonly previousDate: string | undefined;
  readonly forwardLinkAdded: boolean;
}

export interface UpdateReflectionResult {
  readonly reflectionCreated: boolean;
  readonly sectionCreated: boolean;
  readonly linkInserted: boolean;
  readonly forwardLinkAdded: boolean;
  readonly previousDate: string | undefined;
}

// --- Pure Functions (no I/O) ---

export const buildReflectionContent = (date: string, previousDate?: string): string => {
  const backLink = previousDate ? ` | [[reflections/${previousDate}|⏮️]]` : "";
  return `---
share: true
aliases:
  - ${date}
title: ${date}
URL: https://bagrounds.org/reflections/${date}
Author: "[[bryan-grounds]]"
tags:
---
[[index|Home]] > [[reflections/index|Reflections]]${backLink}
# ${date}
`;
};

export const buildSeriesSectionHeading = (series: BlogSeriesConfig): string =>
  `## [[${series.id}/index|${series.icon} ${series.name}]]`;

export const buildPostLink = (seriesId: string, filenameNoExt: string, displayTitle: string): string =>
  `- [[${seriesId}/${filenameNoExt}|${displayTitle}]]`;

export const addForwardLink = (content: string, targetDate: string): string => {
  const forwardLink = `[[reflections/${targetDate}|⏭️]]`;
  return content.includes("⏭️") ? content : content.replace("⏮️]]", `⏮️]] ${forwardLink}`);
};

const findFirstEmbedSectionIndex = (content: string): number => {
  const indices = EMBED_SECTION_HEADERS
    .map((header) => content.indexOf(header))
    .filter((index) => index >= 0);
  return indices.length > 0 ? Math.min(...indices) : -1;
};

const appendLinkToExistingSection = (content: string, sectionHeading: string, postLink: string): string => {
  const lines = content.split("\n");
  const sectionIndex = lines.findIndex((line) => line.startsWith(sectionHeading));
  let insertAt = sectionIndex + 1;
  while (insertAt < lines.length && lines[insertAt]!.startsWith("- ")) {
    insertAt++;
  }
  return [...lines.slice(0, insertAt), postLink, ...lines.slice(insertAt)].join("\n");
};

const insertNewSection = (content: string, sectionHeading: string, postLink: string): string => {
  const firstEmbedIndex = findFirstEmbedSectionIndex(content);
  const sectionBlock = `${sectionHeading}\n${postLink}`;

  return firstEmbedIndex >= 0
    ? `${content.slice(0, firstEmbedIndex).trimEnd()}\n\n${sectionBlock}\n\n${content.slice(firstEmbedIndex)}`
    : `${content.trimEnd()}\n\n${sectionBlock}\n`;
};

export const insertPostLink = (
  content: string,
  series: BlogSeriesConfig,
  filenameNoExt: string,
  displayTitle: string,
): string => {
  const linkTarget = `[[${series.id}/${filenameNoExt}|`;
  if (content.includes(linkTarget)) return content;

  const sectionHeading = buildSeriesSectionHeading(series);
  const postLink = buildPostLink(series.id, filenameNoExt, displayTitle);

  return content.includes(sectionHeading)
    ? appendLinkToExistingSection(content, sectionHeading, postLink)
    : insertNewSection(content, sectionHeading, postLink);
};

// --- I/O Functions ---

export const findPreviousReflectionDate = (reflectionsDir: string, today: string): string | undefined => {
  if (!fs.existsSync(reflectionsDir)) return undefined;

  const candidates = fs.readdirSync(reflectionsDir)
    .filter((f) => /^\d{4}-\d{2}-\d{2}\.md$/.test(f) && f < `${today}.md`)
    .sort();

  const latest = candidates.at(-1);
  return latest?.replace(/\.md$/, "");
};

export const ensureDailyReflection = (reflectionsDir: string, today: string): EnsureReflectionResult => {
  const reflectionPath = path.join(reflectionsDir, `${today}.md`);

  if (fs.existsSync(reflectionPath)) {
    return { created: false, previousDate: undefined, forwardLinkAdded: false };
  }

  const previousDate = findPreviousReflectionDate(reflectionsDir, today);
  const content = buildReflectionContent(today, previousDate);
  fs.mkdirSync(reflectionsDir, { recursive: true });
  fs.writeFileSync(reflectionPath, content, "utf-8");

  let forwardLinkAdded = false;
  if (previousDate) {
    const prevPath = path.join(reflectionsDir, `${previousDate}.md`);
    if (fs.existsSync(prevPath)) {
      const prevContent = fs.readFileSync(prevPath, "utf-8");
      const updated = addForwardLink(prevContent, today);
      if (updated !== prevContent) {
        fs.writeFileSync(prevPath, updated, "utf-8");
        forwardLinkAdded = true;
      }
    }
  }

  return { created: true, previousDate, forwardLinkAdded };
};

export const updateDailyReflection = (
  vaultDir: string,
  today: string,
  series: BlogSeriesConfig,
  postFilename: string,
  postTitle: string,
): UpdateReflectionResult => {
  const reflectionsDir = path.join(vaultDir, "reflections");
  const { created, previousDate, forwardLinkAdded } = ensureDailyReflection(reflectionsDir, today);

  const reflectionPath = path.join(reflectionsDir, `${today}.md`);
  const content = fs.readFileSync(reflectionPath, "utf-8");
  const filenameNoExt = postFilename.replace(/\.md$/, "");
  const hadSection = content.includes(buildSeriesSectionHeading(series));
  const updated = insertPostLink(content, series, filenameNoExt, postTitle);

  const linkInserted = updated !== content;
  if (linkInserted) {
    fs.writeFileSync(reflectionPath, updated, "utf-8");
  }

  return {
    reflectionCreated: created,
    sectionCreated: !hadSection && linkInserted,
    linkInserted,
    forwardLinkAdded,
    previousDate,
  };
};

File: haskell/src/Automation/DailyReflection.hs
module Automation.DailyReflection
  ( UpdateReflectionResult (..)
  , buildReflectionContent
  , buildSeriesSectionHeading
  , buildPostLink
  , addForwardLink
  , insertPostLink
  , findPreviousReflectionDate
  , ensureDailyReflection
  , updateDailyReflection
  , findFirstSectionIndex
  , embedSectionHeaders
  , changesLink
  , ChangesStats (..)
  , renderChangesStats
  , buildChangesStatsPreview
  , upsertChangesPreview
  ) where

import Control.Monad (when)
import Data.List (sort)
import Data.Maybe (catMaybes)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Time (Day)
import System.Directory (createDirectoryIfMissing, doesDirectoryExist, doesFileExist, listDirectory)
import System.FilePath ((</>), dropExtension)

import Automation.Frontmatter (quoteYamlValue)
import Automation.PacificTime (formatDay)

import Automation.BlogSeriesConfig (BlogSeriesConfig (..))
import Automation.Title (Title, unTitle)
import Automation.Wikilink (formatWikilink, NavigableDirectory (..), directoryIndexLink, buildNavBackLink, insertForwardNavLink)
import qualified Automation.Platforms.Bluesky as Bluesky
import qualified Automation.Platforms.Mastodon as Mastodon
import qualified Automation.Platforms.Twitter as Twitter
import Automation.Platform (updatesSectionHeader)
import qualified Automation.DailyReflection.EnsureResult as EnsureResult

embedSectionHeaders :: [Text]
embedSectionHeaders = [Twitter.sectionHeader, Bluesky.sectionHeader, Mastodon.sectionHeader]

changesLink :: Text
changesLink = "## " <> formatWikilink "changes/index" "\128260 Changes"

-- | Statistics for a day's changes page. Using structured integer fields rather than
-- pre-rendered text ensures these values are constructed at the source (from live PageEntry
-- data) and rendered only through 'renderChangesStats', keeping presentation separate from data.
data ChangesStats = ChangesStats
  { statsPageCount     :: Int
  , statsImageCount    :: Int
  , statsLinkCount     :: Int
  , statsBlueskyCount  :: Int
  , statsMastodonCount :: Int
  , statsTwitterCount  :: Int
  } deriving (Show, Eq)

renderChangesStats :: ChangesStats -> Text
renderChangesStats stats =
  let pageWord = if statsPageCount stats == 1 then "page" else "pages"
      counts = catMaybes
        [ renderStatCount (statsImageCount stats) "🖼️" "images"
        , renderStatCount (statsLinkCount stats) "🔗" "links"
        , renderStatCount (statsBlueskyCount stats) "🦋" "Bluesky"
        , renderStatCount (statsMastodonCount stats) "🐘" "Mastodon"
        , renderStatCount (statsTwitterCount stats) "🐦" "Twitter"
        ]
      parts = (T.pack (show (statsPageCount stats)) <> " " <> pageWord) : counts
  in "📊 " <> T.intercalate " · " parts

renderStatCount :: Int -> Text -> Text -> Maybe Text
renderStatCount count emoji label
  | count > 0 = Just (T.pack (show count) <> " " <> emoji <> " " <> label)
  | otherwise = Nothing

changesStatsPreviewLinePrefix :: Text
changesStatsPreviewLinePrefix = "[[changes/"

buildChangesStatsPreview :: Day -> ChangesStats -> Text
buildChangesStatsPreview date stats =
  formatWikilink ("changes/" <> formatDay date) (formatDay date) <> " | " <> renderChangesStats stats

upsertChangesPreview :: Text -> Day -> ChangesStats -> Text
upsertChangesPreview content date stats =
  let preview = buildChangesStatsPreview date stats
  in if T.isInfixOf changesLink content
    then replaceChangesSection content preview
    else T.stripEnd content <> "\n\n" <> changesLink <> "\n" <> preview <> "\n"

replaceChangesSection :: Text -> Text -> Text
replaceChangesSection content preview =
  let contentLines = T.splitOn "\n" content
      (before, fromChanges) = break (== changesLink) contentLines
  in case fromChanges of
    [] -> content
    (_ : rest) ->
      let afterOld = case rest of
            (nextLine : remaining)
              | T.isPrefixOf changesStatsPreviewLinePrefix nextLine -> remaining
            _ -> rest
      in T.intercalate "\n" (before <> [changesLink, preview] <> afterOld)

trailingSectionHeaders :: [Text]
trailingSectionHeaders = updatesSectionHeader : changesLink : embedSectionHeaders

data UpdateReflectionResult = UpdateReflectionResult
  { reflectionCreated :: Bool
  , sectionCreated    :: Bool
  , linkInserted      :: Bool
  , forwardLinkAdded  :: Bool
  , previousDate      :: Maybe Text
  } deriving (Show, Eq)

buildReflectionContent :: Day -> Maybe Text -> Text
buildReflectionContent date previousDate =
  let dateText = formatDay date
      backLink = maybe "" (\pd -> " | " <> buildNavBackLink Reflections pd) previousDate
  in T.intercalate "\n"
    [ "---"
    , "share: true"
    , "aliases:"
    , "  - " <> quoteYamlValue dateText
    , "title: " <> quoteYamlValue dateText
    , "URL: " <> quoteYamlValue ("https://bagrounds.org/reflections/" <> dateText)
    , "Author: \"[[bryan-grounds]]\""
    , "tags:"
    , "---"
    , formatWikilink "index" "Home" <> " > " <> directoryIndexLink Reflections <> backLink
    , "# " <> dateText
    , ""
    , changesLink
    , ""
    ]

buildSeriesSectionHeading :: BlogSeriesConfig -> Text
buildSeriesSectionHeading series =
  "## " <> formatWikilink (seriesId series <> "/index") (icon series <> " " <> name series)

buildPostLink :: Text -> Text -> Title -> Text
buildPostLink seriesId filenameNoExt displayTitle =
  "- " <> formatWikilink (seriesId <> "/" <> filenameNoExt) (unTitle displayTitle)

addForwardLink :: Text -> Text -> Text
addForwardLink = insertForwardNavLink Reflections

findFirstSectionIndex :: [Text] -> Text -> Maybe Int
findFirstSectionIndex headers content =
  case filter (`T.isInfixOf` content) headers of
    [] -> Nothing
    found ->
      let positions = fmap (\h -> T.length $ fst $ T.breakOn h content) found
      in Just $ minimum positions

appendLinkToExistingSection :: Text -> Text -> Text -> Text
appendLinkToExistingSection content sectionHeading postLink =
  let ls = T.splitOn "\n" content
      sectionIndex = length $ takeWhile (not . T.isPrefixOf sectionHeading) ls
      (before, after) = splitAt (sectionIndex + 1) ls
      (listItems, rest) = span (T.isPrefixOf "- ") after
  in T.intercalate "\n" (before <> listItems <> [postLink] <> rest)

insertNewSection :: Text -> Text -> Text -> Text
insertNewSection content sectionHeading postLink =
  let sectionBlock = sectionHeading <> "\n" <> postLink
  in case findFirstSectionIndex trailingSectionHeaders content of
    Just idx ->
      let (before, after) = T.splitAt idx content
      in T.stripEnd before <> "\n\n" <> sectionBlock <> "\n\n" <> after
    Nothing ->
      T.stripEnd content <> "\n\n" <> sectionBlock <> "\n"

insertPostLink :: Text -> BlogSeriesConfig -> Text -> Title -> Maybe Text -> Text
insertPostLink content series filenameNoExt displayTitle replacingFilenameNoExt =
  let linkTarget = "[[" <> seriesId series <> "/" <> filenameNoExt <> "|"
  in if T.isInfixOf linkTarget content
    then content
    else
      let postLink = buildPostLink (seriesId series) filenameNoExt displayTitle
          replacedContent = case replacingFilenameNoExt of
            Just oldName ->
              let oldLinkTarget = "- [[" <> seriesId series <> "/" <> oldName <> "|"
              in if T.isInfixOf oldLinkTarget content
                then
                  let ls = T.splitOn "\n" content
                  in T.intercalate "\n" $ fmap (\l -> if T.isPrefixOf oldLinkTarget l then postLink else l) ls
                else content
            Nothing -> content
          sectionHeading = buildSeriesSectionHeading series
      in if T.isInfixOf sectionHeading replacedContent
        then appendLinkToExistingSection replacedContent sectionHeading postLink
        else insertNewSection replacedContent sectionHeading postLink

isDateFile :: Text -> Text -> Bool
isDateFile today f =
  T.length f >= 13
    && T.isSuffixOf ".md" f
    && T.index f 4 == '-'
    && T.index f 7 == '-'
    && f < (today <> ".md")

findPreviousReflectionDate :: FilePath -> Text -> IO (Maybe Text)
findPreviousReflectionDate reflectionsDir today = do
  exists <- doesDirectoryExist reflectionsDir
  if exists
    then do
      entries <- listDirectory reflectionsDir
      let candidates = sort $ filter (isDateFile today) $ fmap T.pack entries
      pure $ case candidates of
        [] -> Nothing
        _  -> Just $ T.pack $ dropExtension $ T.unpack $ last candidates
    else pure Nothing

ensureDailyReflection :: FilePath -> Day -> IO EnsureResult.EnsureReflectionResult
ensureDailyReflection reflectionsDir today = do
  let todayText = formatDay today
      reflectionPath = reflectionsDir </> T.unpack todayText <> ".md"
  exists <- doesFileExist reflectionPath
  if exists
    then pure EnsureResult.EnsureReflectionResult
      { EnsureResult.reflectionCreated = False, EnsureResult.previousDate = Nothing, EnsureResult.forwardLinkAdded = False }
    else do
      maybePreviousDate <- findPreviousReflectionDate reflectionsDir todayText
      let content = buildReflectionContent today maybePreviousDate
      createDirectoryIfMissing True reflectionsDir
      TIO.writeFile reflectionPath content
      didAddForwardLink <- case maybePreviousDate of
        Nothing -> pure False
        Just pd -> do
          let prevPath = reflectionsDir </> T.unpack pd <> ".md"
          prevExists <- doesFileExist prevPath
          if prevExists
            then do
              prevContent <- TIO.readFile prevPath
              let updated = addForwardLink prevContent todayText
              if updated == prevContent
                then pure False
                else TIO.writeFile prevPath updated >> pure True
            else pure False
      pure EnsureResult.EnsureReflectionResult
        { EnsureResult.reflectionCreated = True, EnsureResult.previousDate = maybePreviousDate, EnsureResult.forwardLinkAdded = didAddForwardLink }

updateDailyReflection :: FilePath -> Day -> BlogSeriesConfig -> Text -> Title -> Maybe Text -> IO UpdateReflectionResult
updateDailyReflection vaultDir today series postFilename postTitle replacingFilename = do
  let todayText = formatDay today
      reflectionsDir = vaultDir </> "reflections"
  EnsureResult.EnsureReflectionResult{..} <- ensureDailyReflection reflectionsDir today
  let reflectionPath = reflectionsDir </> T.unpack todayText <> ".md"
  content <- TIO.readFile reflectionPath
  let filenameNoExt = T.pack $ dropExtension $ T.unpack postFilename
      replacingFilenameNoExt = fmap (T.pack . dropExtension . T.unpack) replacingFilename
      hadSection = T.isInfixOf (buildSeriesSectionHeading series) content
      updated = insertPostLink content series filenameNoExt postTitle replacingFilenameNoExt
      linkInserted' = updated /= content
  when linkInserted' $
    TIO.writeFile reflectionPath updated
  pure UpdateReflectionResult
    { reflectionCreated = reflectionCreated
    , sectionCreated    = not hadSection && linkInserted'
    , linkInserted      = linkInserted'
    , forwardLinkAdded  = forwardLinkAdded
    , previousDate      = previousDate
    }
